### DISCUSSION TITLE
Battery Health Issue with Model S

### DISCUSSION THREAD

**Discussion by jensk22** [teslamate-org/teslamate#3470] [General]
I just did a pull+compose and saw the new Battery Healt Dashboard. On My Tesla Model S 70D which has a 71,2 Brutto, 68,8 Netto and later 67,2 kWh Usable battery, the 'Teslamate Caluclated Max Capacity' is way of. It is 83,7kWh for my model. The calculated Rated Range efficiency is caklculated to 150Wh/km and the Typical is calculated to 190Wh/km. I could guess that there is a hickup here, because if I multiply 83,7kWh * 150Wh/km / 190Wh/km I get an almost reasonable Original capacity of 66,1 kWh. 

Something is definately wrong, here is my 'All Logged Time' capacity:

![image](https://github.com/teslamate-org/teslamate/assets/92438498/be395ba4-7b25-4974-855d-64aa35d9a3cc)

The calculated Usable of 65kWh is spot on with ScanMyTesla, which reports 65,1kWh (that is with the newer 4kWh buffer, as opposed to the 2,4kWh buffer my car originally had according to Jason Hughed)

(I HAVE done a search for simlar discussions and did find an 85D owner complaining, that the +90kWh Original Estimate was dubious, but I can no longer find that issue.)

**Comment by ktsenvolt** [comment]
I am having some similar issues with #3458 for the 2020 M3 SR+.
I don't know what is the difference between ODB PID (Scan my Tesla) and Tesla API data.
[@juanheredia](https://twitter.com/JuanHeredia) should know.

**Comment by jensk22** [comment]
I tried to change the Total Capacity Graph Query to use the charges.ideal_battery_range_km (instead of the charges.rated_battery_range_km) and that produces proper 'old' capacity numbers (which for me is around 66.1kWh because I started using Teslamata at odeometer=70kKm). So the : 

('$aux'::json -> 'RatedEfficiency')::text::float 

Is definately not the Rated Efficiency of 150Wh/km, but the Ideal Efficiency which is 190Wh/km.

SELECT convert_km(AVG(p.odometer)::numeric,'$length_unit') AS odometer, 
	AVG(c.ideal_battery_range_km * ('$aux'::json -> 'RatedEfficiency')::text::float / c.usable_battery_level) AS "kWh",
	MAX(cp.id) AS id,
	to_char(cp.end_date, 'YYYY-MM-dd') AS Title
	FROM charging_processes cp
		JOIN (SELECT charging_process_id, MAX(date) as date	FROM charges WHERE usable_battery_level > 0 GROUP BY charging_process_id) AS last_charges	ON cp.id = last_charges.charging_process_id
		INNER JOIN charges c
		ON c.charging_process_id = cp.id AND c.date = last_charges.date
		INNER JOIN positions p ON p.id = cp.position_id
	WHERE cp.car_id = $car_id
		AND cp.end_date IS NOT NULL
		AND cp.charge_energy_added >= ('$aux'::json -> 'RatedEfficiency')::text::float
	GROUP BY 4
	
	
![image](https://github.com/teslamate-org/teslamate/assets/92438498/04596d4a-9f0f-4ad4-8f15-014a0643cde2)

**Comment by jheredianet** [comment]
> I just did a pull+compose and saw the new Battery Healt Dashboard. On My Tesla Model S 70D which has a 71,2 Brutto, 68,8 Netto and later 67,2 kWh Usable battery, the 'Teslamate Caluclated Max Capacity' is way of. It is 83,7kWh for my model. The calculated Rated Range efficiency is caklculated to 150Wh/km and the Typical is calculated to 190Wh/km. I could guess that there is a hickup here, because if I multiply 83,7kWh * 150Wh/km / 190Wh/km I get an almost reasonable Original capacity of 66,1 kWh.

> The calculated Usable of 65kWh is spot on with ScanMyTesla, which reports 65,1kWh (that is with the newer 4kWh buffer, as opposed to the 2,4kWh buffer my car originally had according to Jason Hughed)

Hi, estimation of the battery capacity (when new) is based on the range looged on the database, it's not a  a reading from ODB (like ScanMy Tesla and other tools do). So, I guess in some moment your max rated_battery_range_km (436.7km in your case) was misreported by Tesla API or you have imported data from other systems.

In the other hand, maybe you want to browse your database and check if there is some charge session or drive not closed. 

This would be the query you could use to get the max range reported by the car (assuming your car_id is 1, if not change it accordantly)
````
SELECT 
  floor(extract(epoch from date)/86400)*86400 AS time, 
  sum(rated_battery_range_km) / sum(usable_battery_level) * 100 AS max_rated_range_km 
 FROM ( 
  SELECT battery_level, usable_battery_level, date,  rated_battery_range_km from charges c  
  JOIN charging_processes p ON p.id = c.charging_process_id  
  WHERE p.car_id = 1 AND usable_battery_level IS NOT NULL) AS data 
 GROUP BY 1 
 ORDER BY 2 DESC 
 LIMIT 1
````
For efficiency, you could use this query:

```
SELECT efficiency * 100.0 AS rated_efficiency FROM cars WHERE id = 1
````

**Reply by jheredianet** [reply]
Sorry, I wouldn't know how to make a calculation that only works for your car or specifically for that model. I don't know which models have this peculiarity. The data is extracted from what is stored in the "rated_efficiency" field of the "cars" table. Something you could do is to modify the value you have in that table and test if it adapts to what you mention, but I don't know what other implications it could have with other calculations from other dashboards.

**Comment by jheredianet** [comment]
> I tried to change the Total Capacity Graph Query to use the charges.ideal_battery_range_km (instead of the charges.rated_battery_range_km) and that produces proper 'old' capacity numbers (which for me is around 66.1kWh because I started using Teslamata at odeometer=70kKm). 

Hi, check my previous comment. 
Remember that these estimations are based on the data logged in Teslamate. You may check/browse your data step by step in other to get which record is fouling the sample. 
About Ideal and rated range, IMHO, it's better to do the math based on the rated one because all the data reported by the API takes as a reference the car's efficiency defined by Tesla, without considering your driving habits or type of roads, terrain, climate, etc. If you use the ideal range, you will see that the degradation points will be very diverse and will not appear accordantly.

As you can see in the graph "hint" 
**Usable (new)** is the estimated Battery Capacity since you began to use Teslamate. That's why, the more data you have logged from your brand-new car the better. For those who have not used Teslamate since they got their new car, or for those who have bought it second hand,  it's possible to set the max range to 100% and the battery capacity of the car battery when it was new in order to get a better and accurate estimation.

**Comment by jensk22** [comment]
Hi Juan

I fear you are over complicating things !

I totally agree you shoul duse Rated Efficiency, (which I as well wrote in my Issue report!)

I specifically propose that the Rated Efficiency be corrected, because using Ideal is error prone (as below documents, Tesla Chenged the Ideal Range Wh/km here on May 9 in April 2016!

I have used another CloudsService for years and it shows that Ideal Efficiency was changed from approx 187Wh/km to 190Wh/km causing below drop in 100% Range![cid:e18b62df-3bee-411f-8b0a-a8c54fca0995]

As you assume, the Rated Efficiency has been less changed, so my Rated Range Decradation is smooth as silk:
[cid:2efb4625-c235-40cb-a54f-58debccacd1a]

As Teslamate does, I originally (in 2016 - 2020) as well calculated the Rated Wh/km, and it has not been totally fixed, but less modified (below graph shows all Outliers no matter how small the charges was 🙁

[cid:69a9b077-3743-496a-bea6-f1733f372bd9]
So I just want you to use the registered Rated efficiency of 150Wh/km instead of the 190Wh/km, that is currently used for my car. That is all 🙂

PS: As to open charges to explain the problem, I think that is unlikely to be the issue, first of all because I fixed both Open Drives and Open Charges a year ago and they have not shows up in later versions of Teslamate. And your graph shows that ALL the reported charges concerted to 100% is way off and above 81 kWh.
[cid:76316853-8765-4a2b-8e17-33afe1828bfb]

HTH

best regards
Jens

________________________________

**Comment by jensk22** [comment]
Hi Juan

I looked at the Teslamate tables using 
    teslamate=# select * from information_schema.tables where table_schema = 'public';

and decided that the charging_processes has kWh added as well as rated- and and ideal_km_added, so I propose that Rated and Ideal Efficiency can be calculated as:
    teslamate=# select charge_energy_added, end_rated_range_km - start_rated_range_km as RatedKm, end_ideal_range_km - start_ideal_range_km as IdealKm, charge_energy_added / (end_rated_range_km - start_rated_range_km) as RatedkWhKm, charge_energy_added / (end_ideal_range_km - start_ideal_range_km) as IdealkWhPerKm  from charging_processes where charge_energy_added > 10;

 charge_energy_added | ratedkm | idealkm |       ratedkwhkm       |     idealkwhperkm
---------------------+---------+---------+------------------------+------------------------
               10.34 |   68.91 |   54.40 | 0.15005079088666376433 | 0.19007352941176470588
               11.94 |   79.61 |   62.86 | 0.14998115814596156262 | 0.18994591154947502386
               10.24 |   68.24 |   53.89 | 0.15005861664712778429 | 0.19001670068658378178
               12.26 |   81.69 |   64.50 | 0.15007956910270534949 | 0.19007751937984496124
               25.04 |  166.92 |  131.79 | 0.15001198178768272226 | 0.18999924121708779118
               23.44 |  156.25 |  123.36 | 0.15001600000000000000 | 0.19001297016861219196
               14.59 |   97.30 |   76.80 | 0.14994861253854059609 | 0.18997395833333333333
               13.96 |   93.79 |   74.05 | 0.14884316025162597292 | 0.18852126941255908170
               43.60 |  290.63 |  229.44 | 0.15001892440560162406 | 0.19002789400278940028
               14.51 |   96.73 |   76.35 | 0.15000516902718908301 | 0.19004584151931892600
               14.28 |   95.19 |   75.16 | 0.15001575795776867318 | 0.18999467802022352315
               10.76 |   72.47 |   57.22 | 0.14847523113012280944 | 0.18804613771408598392
               13.64 |   90.91 |   71.77 | 0.15003849961500384996 | 0.19005155357391667828
               10.23 |   68.18 |   53.82 | 0.15004400117336462306 | 0.19007803790412486065
               10.98 |   73.16 |   57.76 | 0.15008201202843083652 | 0.19009695290858725762
               19.32 |  128.78 |  101.66 | 0.15002329554278614692 | 0.19004524886877828054
               16.32 |  108.78 |   85.88 | 0.15002757859900717044 | 0.19003260363297624592
               10.34 |   68.92 |   54.41 | 0.15002901915264074289 | 0.19003859584635177357
               16.20 |  107.99 |   85.25 | 0.15001389017501620520 | 0.19002932551319648094
               30.90 |  206.04 |  162.66 | 0.14997087944088526500 | 0.18996680191811139801
               18.12 |  120.81 |   95.39 | 0.14998758380928731065 | 0.18995701855540413041
               33.68 |  224.54 |  177.27 | 0.14999554645052106529 | 0.18999266655384441812
               23.03 |  153.50 |  121.18 | 0.15003257328990228013 | 0.19004786268361115696
               13.75 |   92.38 |   72.95 | 0.14884174063650140723 | 0.18848526387936943112
               11.83 |   79.59 |   62.83 | 0.14863676341248900616 | 0.18828585070826038517
               34.98 |  230.23 |  181.78 | 0.15193502150023889154 | 0.19243041038618109803
               18.54 |  123.63 |   97.59 | 0.14996360106770201407 | 0.18997848140178296957
--More--


I don't know how Adrian calculates the efficiencies, but given Teslamate is open source a more 'into the code' guy shoul dbe able to decipher 🙂
But they are shown here:

![image](https://github.com/teslamate-org/teslamate/assets/92438498/c4a968a2-6f14-4c0a-8319-a8289814816c)


regards
Jens

**Reply by jheredianet** [reply]
Oh, I see. So, you are suggesting that the dashboard should consider the derivated rated efficiencies. No the cuestion is, which one? I suggest that one that has more counts. That's why I've decided to consider only the Current rated Efficency (setup by Tesla). 

For example, in my case (M3 SR+ 2020 Europe), the efficiency is always (almost) the same, that's why i just use the current rated efficiency.

![image](https://github.com/teslamate-org/teslamate/assets/45735191/1d85a088-0eaa-4306-8662-5337e03701e5)


Let me dig a little more and try to make some tests, so I may do a pull request to update the dashboard.

**Comment by ktsenvolt** [comment]
![image](https://github.com/teslamate-org/teslamate/assets/80360674/ec2a74a5-330b-4bc6-af18-480d5732603b)
My 2020 M3 SR+ worked well on the efficiency. Maybe there is a difference between M3 and MS when pulling the data API.

**Comment by jensk22** [comment]
Hi Juan

My databse has only one %efficieny% and it is th eone in Cars:
teslamate=# select table_name, column_name from information_schema.columns where column_name like '%efficiency%';
 table_name | column_name
------------+-------------
 cars       | efficiency
(1 row)

And the value is:
teslamate=# select efficiency from cars;
 efficiency
------------
    0.18997
(1 row)

I guess that to handle older Model S' you would need:
A. Calulate Rated efficiency yourself
or
B: Use Ideal (if Ideal is as well available on M3 and MY?)
C: Best: Add rated_efficiency to the cars table :-) 

best regards
jens

PS: I have just discoverede that email diskucssions DON'T insert pictures, so here are the Rated versus ideal graphs from my 'old' logger + Excel!

Ideal effieciency was altered on May 9 2016 for my Model S 70D:
![image](https://github.com/teslamate-org/teslamate/assets/92438498/c021c032-1e3b-4c76-9317-cf6b2a82623e)

Rated Efficiency has been almost unchanged, guess it started out at 148Wh/km, then went stable at 150Wh/km:
![image](https://github.com/teslamate-org/teslamate/assets/92438498/9214379b-9f39-4b9e-85a4-888ebf863f8d)

**Comment by jensk22** [comment]
I guess ?Adrian? could have named a potential 'rated wh' column something else than 'efficiency', so I tried below with no luck :-(

teslamate=# select table_name, column_name from information_schema.columns where column_name like '%rated%';
     table_name     |      column_name
--------------------+------------------------
 charges            | rated_battery_range_km
 pg_attribute       | attgenerated
 charging_processes | end_rated_range_km
 charging_processes | start_rated_range_km
 drives             | start_rated_range_km
 drives             | end_rated_range_km
 positions          | rated_battery_range_km
 columns            | is_generated
(8 rows)

teslamate=# select table_name, column_name from information_schema.columns where column_name like '%wh%';
       table_name        | column_name
-------------------------+-------------
 pg_foreign_data_wrapper | fdwhandler
 cars                    | wheel_type
(2 rows)

**Comment by jheredianet** [comment]
I'll try to make the changes of the dashboard on my own repository first https://github.com/jheredianet/Teslamate-CustomGrafanaDashboards

So, many other users can test and give feedback before making a PR on Teslamate. 
I'll let you know when it is done.

**Reply by jheredianet** [reply]
Hi @jensk22 

I've made un update for your suggestion. It's on my own repository ("on develop branch") at https://github.com/jheredianet/Teslamate-CustomGrafanaDashboards/tree/develop

You may import manually the BatteryHealth dashboard as explained there and test if the estimations now are more accurate. If so, I'll made a pull request to be used in Teslamate. 

Let me know how it goes.

**Reply by jensk22** [reply]
I have tested and confirmed that your fix work like a charm and have confirmed in Issue #3486.
@jensk22
Author
[jensk22](https://github.com/jensk22) commented [on Dec 1, 2023](https://github.com/teslamate-org/teslamate/issues/3486#issuecomment-1835718282)
I Imported the JSON Content as 'Battery Health2' and results are fine!
![image](https://github.com/teslamate-org/teslamate/assets/92438498/63cf625f-7d0a-4ca4-89bb-eac4be90bbee)

But that issue has been closed as not planned, so I guess your fix will not be released :-(
